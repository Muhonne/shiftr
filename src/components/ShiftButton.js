// @flow

import React from "react";
import { ActivityIndicator, LayoutAnimation } from "react-native";
import styled from "styled-components";
import { reduxAutoloader } from "redux-autoloader";

import apiCalls from "../apiCalls";
import Button from "./Button";
import Text from "./Text";
import constants from "../constants";

const View = styled.View``;

const ButtonContent = styled.View`
  padding: ${constants.spacing.s}px;
  border-width: 1;
  border-color: ${constants.colors.darkGreen};
  border-radius: ${constants.spacing.s};
  align-items: center;
  justify-content: center;
  flex-direction: row;
  background-color: ${props =>
    props.error ? constants.colors.red : constants.colors.white};
`;

type Props = {
  shiftId: string, // eslint-disable-line react/no-unused-prop-types,:type used in connect
  isLoading: boolean,
  data: any,
  refresh: () => void,
  refreshParent: () => void,
  parentBooked: boolean,
  label: string
};

const ShiftButton = (props: Props) => {
  LayoutAnimation.easeInEaseOut();
  const {
    refresh,
    data,
    isLoading,
    refreshParent,
    parentBooked,
    label
  } = props;
  if (data && !data.statusCode && data.booked !== parentBooked) {
    refreshParent();
  }
  return (
    <View>
      <Button onPress={refresh}>
        <ButtonContent>
          <Text large>{label}</Text>
          {isLoading && <ActivityIndicator size="small" />}
        </ButtonContent>
      </Button>
      {data && data.message && <Text>{data.message}</Text>}
    </View>
  );
};

export default reduxAutoloader({
  startOnMount: false,
  loadOnInitialize: false,
  reloadOnMount: false,
  name: (props: Props) => `${props.shiftId}.status`,
  apiCall: (props: Props) => apiCalls[props.label.toLowerCase()](props.shiftId)
})(ShiftButton);
