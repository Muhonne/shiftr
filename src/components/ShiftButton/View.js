// @flow

import React from "react";
import styled from "styled-components";
import Button from "../Button";
import LayoutAnimation from "../../LayoutAnimation";
import ActivityIndicator from "../ActivityIndicator";
import Text from "../Text";
import constants from "../../constants";

const View = styled.View``;

const ButtonContent = styled.View`
  padding: ${constants.spacing.s}px;
  margin-bottom: ${constants.spacing.s};
  border-width: 1;
  border-color: ${constants.colors.woltishBlue};
  align-items: center;
  justify-content: center;
  flex-direction: row;
  opacity: ${props => (props.error ? 0.5 : 1)};
`;

type Props = {
  isLoading: boolean,
  data: any,
  refresh: () => void,
  refreshParent: () => void,
  parentBooked: boolean,
};

const ShiftButton = (props: Props) => {
  LayoutAnimation();
  const {
    refresh,
    data,
    isLoading,
    refreshParent,
    parentBooked,
  } = props;
  if (data && !data.statusCode && data.booked !== parentBooked) {
    // if it's stupid but it works it's not stupid
    // regardless this is a bit shit and there is no refreshing/loading indicator on the parent so
    // if it takes a while to update the text won't change and it looks like nothing is happening
    refreshParent();
  }
  const error = data && data.message;
  return (
    <View>
      <Button onPress={refresh}>
        <ButtonContent error={error}>
          <Text
            large
            style={{
              color: constants.colors.woltishBlue
            }}
          >
            {parentBooked ? "Cancel" : "Book"}
          </Text>
          {isLoading && (
            <ActivityIndicator style={{ marginLeft: 5 }} size="small" />
          )}
        </ButtonContent>
      </Button>
      {error && <Text center>{data.message}</Text>}
    </View>
  );
};

export default ShiftButton;
