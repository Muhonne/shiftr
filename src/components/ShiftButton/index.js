import { reduxAutoloader } from "redux-autoloader";

import apiCalls from "../../apiCalls";
import View from "./View";

export default reduxAutoloader({
  startOnMount: false,
  loadOnInitialize: false,
  reloadOnMount: false,
  name: props => `${props.shiftId}.status`,
  apiCall: props => apiCalls[props.label.toLowerCase()](props.shiftId)
})(View);
