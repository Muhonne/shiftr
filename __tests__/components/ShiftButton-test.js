import React from "react";
import renderer from "react-test-renderer";
import ShiftButton from "../../src/components/ShiftButton/View";

describe("Button", () => {
  it("renders without props", () => {
    expect(renderer.create(<ShiftButton />)).toMatchSnapshot();
  });

  it("renders a bookable button", () => {
    expect(
      renderer.create(<ShiftButton parentBooked={false} />)
    ).toMatchSnapshot();
  });

  it("renders a cancel button", () => {
    expect(renderer.create(<ShiftButton parentBooked />)).toMatchSnapshot();
  });

  it("renders a button with error", () => {
    expect(
      renderer.create(
        <ShiftButton
          parentBooked
          data={{ message: "test error message" }}
          refreshParent={() => {}}
        />
      )
    ).toMatchSnapshot();
  });
});
