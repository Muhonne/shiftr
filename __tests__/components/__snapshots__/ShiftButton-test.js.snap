// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a bookable button 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    error={undefined}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#006bb4",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 10,
          "opacity": 1,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
        undefined,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      large={true}
      style={
        Array [
          Object {
            "alignSelf": "flex-start",
            "color": "#000",
            "fontSize": 18,
          },
          Object {
            "color": "#006bb4",
          },
        ]
      }
    >
      Book
    </Text>
  </View>
</View>
`;

exports[`Button renders a button with error 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    error="test error message"
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#006bb4",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 10,
          "opacity": 0.5,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
        undefined,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      large={true}
      style={
        Array [
          Object {
            "alignSelf": "flex-start",
            "color": "#000",
            "fontSize": 18,
          },
          Object {
            "color": "#006bb4",
          },
        ]
      }
    >
      Cancel
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    center={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#000",
          "fontSize": 16,
        },
        undefined,
      ]
    }
  >
    test error message
  </Text>
</View>
`;

exports[`Button renders a cancel button 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    error={undefined}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#006bb4",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 10,
          "opacity": 1,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
        undefined,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      large={true}
      style={
        Array [
          Object {
            "alignSelf": "flex-start",
            "color": "#000",
            "fontSize": 18,
          },
          Object {
            "color": "#006bb4",
          },
        ]
      }
    >
      Cancel
    </Text>
  </View>
</View>
`;

exports[`Button renders without props 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    error={undefined}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#006bb4",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 10,
          "opacity": 1,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
        undefined,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      large={true}
      style={
        Array [
          Object {
            "alignSelf": "flex-start",
            "color": "#000",
            "fontSize": 18,
          },
          Object {
            "color": "#006bb4",
          },
        ]
      }
    >
      Book
    </Text>
  </View>
</View>
`;
